python/tank/platform/software_launcher.py

```diff
@@ -239,38 +276,156 @@ def minimum_supported_version(self):
 
-    def scan_software(self, versions=None):
+        Example::
+
+            ["Houdini", "Houdini Core", "Houdini FX"]
         """
-        Performs a scan for software installations.
+        return self._products
 
-        :param list versions: List of strings representing versions
-                              to search for. If set to None or [], search
-                              for all versions. A version string is
-                              DCC-specific but could be something
-                              like "2017", "6.3v7" or "1.2.3.52"
-        :returns: List of :class:`SoftwareVersion` instances
+    @property
+    def versions(self):
         """
-        raise NotImplementedError
+        A list of versions limiting executable discovery.
+
+        Example::
+
+            ["15.5.324", "16.0.1.322"]
+        """
+        return self._versions
+
+    ##########################################################################################
+    # abstract methods
 
     def prepare_launch(self, exec_path, args, file_to_open=None):
         """
         Prepares the given software for launch
 
         :param str exec_path: Path to DCC executable to launch
         :param str args: Command line arguments as strings
-        :param str file_to_open: (optional) Full path name of a file to open on launch
+        :param str file_to_open: (optional) Full path name of a file to open on
+            launch
         :returns: :class:`LaunchInformation` instance
         """
         raise NotImplementedError
 
+    def _scan_software(self):
+        """
+        Abstract method that must be implemented by subclasses to return a list
+        of ``SoftwareVersion`` objects on the local filesystem.
+
+        Typical implementations will use functionality such as ``glob`` to
+        locate all versions and variations of executables on disk and then
+        creating ``SoftwareVersion`` objects for each executable. This method
+        is called by the ``get_supported_software()`` method, and each
+        ``SoftwareVersion`` object returned is checked against the launcher's
+        lists of supported version and product variations via the
+        ``_is_supported`` method.
+        """
+
+        raise NotImplementedError
+
+    def _scan_software_with_expression(self, match_template, regular_expression):
+        """
+        This is a helper method that can be invoked in an implementation of :meth:`_scan_software`.
+
+        The ``match_template`` argument provides a template to use both for globing files and then pattern
+        matching them using regular expressions provided by the ``tokens`` dictionary.
+
+        The method will first substitute every token from the template with a ``*`` for globing files.
+        It will then replace the tokens in the template with the regular expressions that were
+        provided.
+
+        In the following example::
+
+            self._scan_software_with_expression(
```

Comment:
Hum, I was thinking of going from a glob to a regex, which would be difficult because you'd have to distinguish between uses of \\ to escape the character after and when you need to double escape it for a path seperator. but if the interface specifies that you need to provide a regex's template, then we can without a doubt convert \\\\ to \\ in order to build the glob. I'll try it out